question-servers/freeform.js

```diff
@@ -6,10 +6,74 @@ var path = require('path');
 var mustache = require('mustache');
 var cheerio = require('cheerio');
 
-var elements = require('./elements');
+var {
+    elements,
+    dependencies: elementDependencies
+} = require('./elements');
 var codeCaller = require('../lib/code-caller');
 
 module.exports = {
+    /**
+     * Computes the dependencies given a Set of element names.
+     * @param  {Set}      elements
+     * @param  {Function} callback Will be called with an array of <link>/<script> tags
+     */
+    getElementDependencies: function(elements, callback) {
+        const elementStyleFiles = [];
+        const elementScriptFiles = [];
+        for (const elem of elements) {
+            elementStyleFiles.push(`${elem}.css`);
+            elementScriptFiles.push(`${elem}.js`);
+        }
+        const styleUrls = new Set();
+        const scriptUrls = new Set();
+        async.parallel([
+            (callback) => {
+                async.each(elementStyleFiles, (file, callback) => {
+                    fs.access(path.join(__dirname, 'elements', file), (err) => {
+                        if (!err) {
+                            styleUrls.add(`/pl/static/elements/${file}`);
+                            //headerHtmls.push(`<link href="/pl/static/elements/${file}" rel="stylesheet" />`);
+                        }
+                        callback(null);
+                    });
+                }, () => callback(null));
+            },
+            (callback) => {
+                async.each(elementScriptFiles, (file, callback) => {
+                    fs.access(path.join(__dirname, 'elements', file), (err) => {
+                        if (!err) {
+                            scriptUrls.add(`/pl/static/elements/${file}`);
+                            //headerHtmls.push(`<script type="text/javascript" src="/pl/static/elements/${file}"></script>`);
+                        }
+                        callback(null);
+                    });
+                }, () => callback(null));
+            }
+        ], () => {
+            // Now that we've ~automagically~ added all available files, let's
+            // add any other declared dependencies
+            const globalScriptUrls = new Set();
+            for (const elem of elements) {
+                const deps = elementDependencies[elem];
+                if (!deps) continue;
+                deps.styles && deps.styles.forEach((file) => styleUrls.add(`/pl/static/elements/${file}`));
```

Comment:
I don't want to turn this into a full hierarchical dependency resolution engine, if we can possibly avoid doing so :-)

I think in general that this should work like class jquery-type web code, where you have to include all the scripts you want, and if they have dependencies then you need to manually include them earlier in the load order. For example, when using backbone the docs say "please make sure lodash/underscore is already loaded", so you just include lodash and then backbone.

For us, I think the only requirement is that we should be order-preserving in the case of consistent order. I don't think we should detect/error on inconsistent ordering. For example, say there are two unrelated JS libraries A and B, where the load order doesn't matter. Element E1 lists them as A, B while element E2 lists them as B, A. Testing each element individually shows that both elements work fine. But now someone includes both elements within a single question. We don't want to throw a fatal error all of a sudden just because there is apparently a circular dependency.

Of course, one option would be to require that elements specify which things in their dependency list depend on each other. But that's a bit of overkill, I think.

Instead, I think we can just do something like this:

```text
allDependencies = []
for each element on page:
    for each dependency in element.dependencies:
        if !allDependencies.includes(dependency):
            allDependencies.push(dependency)
```

Definition: We say that A is a *dependency* of B if, whenever B appears in an element dependency list, A must also appear and be listed before B.

Theorem: If A is a dependency of B then whenever B appears in allDependencies, A will also occur and will be listed before B.

Proof: Suppose not. That is, B appears in allDependencies without a preceding A. Because each dependency is only pushed onto allDependencies at most once, we can consider the unique element that caused B to be pushed onto allDependencies. This element has B in its dependency list. All dependencies in element.dependency were either already present in allDependencies before B, or were pushed before B. None of these are A, so A does not appear in element.dependencies before B. Thus A is not a dependency of B. Hence proof by contradiction.